transport/src/main/java/io/netty/channel/socket/nio/NioServerSocketChannel.java

```diff
@@ -137,18 +152,34 @@ protected void doClose() throws Exception {
 
     @Override
     protected int doReadMessages(List<Object> buf) throws Exception {
-        SocketChannel ch = javaChannel().accept();
+        if (connectionCounter.incrementAndGet() > maxConnections) {
+            connectionCounter.decrementAndGet();
+            try {
+                //sleep a while so that we won't loop busily
+                Thread.sleep(10);
```

Comment:
I think we will need to use a different strategy here. We should actively shutdown/close the peer socket so:

1. the peer knows the connection has been rejected
2. our local backlog is drained and we can accept more connections in the future if we drop below the "max connections" threshold.

This will also be necessary in the case of Edge Triggered mode for the native transports. In this case we will not be notified by the selector that there is activity on the FD and so we may never accept sockets once we get into this condition.

The risk here is that we exceed the "max limit" temporarily (to accept a FD and then close it). This should be documented as part of the socket option.